src/Lean/Data/Lsp/Basic.lean

```diff
@@ -0,0 +1,313 @@
+/-
+Copyright (c) 2020 Marc Huisinga. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+
+Authors: Marc Huisinga, Wojciech Nawrocki
+-/
+import Lean.Data.Json
+
+/-! Defines most of the 'Basic Structures' in the LSP specification
+(https://microsoft.github.io/language-server-protocol/specifications/specification-current/),
+as well as some utilities.
+
+Since LSP is Json-based, Ints/Nats are represented by Floats on the wire. -/
+
+namespace Lean
+namespace Lsp
+
+open Json
+
+abbrev DocumentUri := String
+
+/-- An array of the lines in a text document.
+Elements mustn't contain newline characters. -/
+def DocumentText := Array String
+
+/-- Can represent both a UTF-16 position - this is what LSP clients send -
+and a codepoint position - what we use internally.
+`character` is accepted liberally: actual character := min(line length, character) -/
+structure Position := (line : Nat) (character : Nat)
+
+instance Position.hasFromJson : HasFromJson Position :=
+⟨fun j => do
+  line ← j.getObjValAs? Nat "line";
+  character ← j.getObjValAs? Nat "character";
+  pure ⟨line, character⟩⟩
+
+instance Position.hasToJson : HasToJson Position :=
+⟨fun o => mkObj [
+  ⟨"line", o.line⟩,
+  ⟨"character", o.character⟩]⟩
+
+instance Position.hasToString : HasToString Position :=
+⟨fun p => "(" ++ toString p.line ++ ", " ++ toString p.character ++ ")"⟩
+
+namespace DocumentText
+
+/-- Computes a linear position in `("\n".intercalate text.toList)`
+from an LSP-style 0-indexed (ln, col) position. -/
+def lnColToLinearPos (text : DocumentText) (pos : Position) : String.Pos :=
+text.foldrRange 0 pos.line (fun ln acc => acc + ln.length + 1) pos.character
+
+/-- An imprecise inverse of lnColToLinearPos.
```

Comment:
It's not a true inverse as `lnColToLinearPos` is not injective. The reason is that we often deal with positions that represent an inserted character, but on a pre-insertion `text` (and similarly for deletions). Such an insertion position at the end of a line has the same linear value as the beginning of the next line. I will update the comment.

Reply:
Well that was a lie. Sorry! The LSP clients never seem to send the kind of position that would break injectivity and there was an off-by-one :sweat_smile: Fixed.